Rewrite Fronius integration - add config_flow, unique_ids, devices

farmio opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
- YAML configuration will be imported to a ConfigEntry
- this will change default entity_ids and might break some customizations, automations and assignments to the energy dashboard for previous users. Since the integration now provides entities with unique_ids it can be changed back to the previous value if desired. 

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a drop-in replacement of the current Fronius integration. 
This integration now
- provides a ConfigFlow
- doesn't need any other configuration than the IP address or hostname of the Fronius device
- provides an import for the now deprecated yaml config
- uses DataUpdateCoordinators
- creates entities with unique_ids
- assigns entities to devices
- uses SensorEntityDescription
- provides nearly all datapoints the old integration had (except for some omitted entities providing timestamp data of results which are always request time +- some milliseconds)
- is fully type annotated


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: 
  - fixes #53827
  - closes #54491 (we can't do anything about it if it is an issue on the API side)
  - closes https://github.com/nielstron/pyfronius/issues/40
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works. 
  - for config_flow

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

farmio committed:
initial refactoring commit - meters

- config_flow (no strings, no tests yet)
- import yaml config
- FroniusSolarNet class for holding Fronius object , coordinators and some common data
- meter descriptions
- update coordinator
- entities (including devices)

farmio committed:
storage controllers

farmio committed:
error handling on init; inverter unique_id

farmio committed:
inverters

farmio committed:
power_flow

farmio committed:
fix VA, var, varh not valid for device_class power/energy

and add custom icons

farmio committed:
add SolarNet device for system wide values

farmio committed:
cleanup

farmio committed:
config_flow strings

farmio committed:
test config_flow

farmio committed:
use pyfronius 0.7.0

farmio committed:
enable strict typing

farmio committed:
remove TODO comments

farmio committed:
fix lint errors; move FroniusSensorEntity to sensor.py

farmio committed:
power_flow as optional coordinator

API V0 doesn't support power_flow endpoint

farmio committed:
show error message in logs

nielstron commented:
I tried the config flow at home and it seems to work perfectly fine. One thing that I did not understand yet: why did you decide to make all Sensor descriptions explicit? Was the pattern matching before too error prone? This way, you need to make sure that all possible sensor values returned by the fronius bridge are caught and displayed.

farmio commented:
I just wanted to be able to set nice default names.
```python
"energy_reactive_ac_consumed": SensorEntityDescription(
        key="energy_reactive_ac_consumed",
        name="Energy Reactive AC consumed",
    )
```
and using this I found that expanding it to a full SensorEntityDescription was a nice way to have everything at one place so it is easy to reason about. This also makes sure proper HA supported unit constants are used (this is probably a bad example as `varh` isn't one of them 🤣).
```python
"energy_reactive_ac_consumed": SensorEntityDescription(
        key="energy_reactive_ac_consumed",
        name="Energy Reactive AC consumed",
        native_unit_of_measurement=ENERGY_VOLT_AMPERE_REACTIVE_HOUR,
        state_class=STATE_CLASS_TOTAL_INCREASING,
        icon="mdi:lightning-bolt-outline",
    )
```
Pattern matching this key for "energy*" would have wrongly set its `device_class=DEVICE_CLASS_ENERGY` leading to log errors
```log
WARNING (Recorder) [homeassistant.components.sensor.recorder] sensor.energy_reactive_ac_consumed_fronius_meter_0_http_10_1_2_5 has unknown unit VArh
```

Having the used keys explicit also enables to use others as `extra_state_attribute` like "Enable" from a meter device. I also deliberately excluded the "timestamp" key as I don't see a point in tracking the time in multiple entities.
![Bildschirmfoto 2021-09-13 um 23 21 12](https://user-images.githubusercontent.com/12422879/133158112-8cd87f7d-9919-401f-9363-c7c428b420d5.png)

nielstron commented:
I just noticed that this drops support for Fronius Logger Info, which is available on most devices and provides useful data like the current energy price. Please do re-add that sensor (or add a note for this in the description)

> Pattern matching this key for "energy*" would have wrongly set its device_class=DEVICE_CLASS_ENERGY leading to log errors

Actually, the unit for reactive energy is not watt-hour precisely so that it is not confused with any actual power production (see https://en.wikipedia.org/wiki/AC_power#Active,_reactive,_apparent,_and_complex_power_in_sinusoidal_steady-state). I think the device class Energy and non-watt-hour units should not necessarily result in a log error.

farmio commented:
> I think the device class Energy and non-watt-hour units should not necessarily result in a log error.

They do. I think because the recorder can't convert the units at some point https://github.com/home-assistant/core/blob/aaa62dadec4ccc019feec797346d412a0a565ace/homeassistant/components/sensor/recorder.py#L82
`DEVICE_CLASS_ENERGY` may only be set for entities with unit Wh or kWh. See also https://developers.home-assistant.io/docs/core/entity/sensor

I deliberately left out Logger info for now because it only adds values that are constants which have to be set in the inverters Web-UI by the user. I think the least users have even set this, let alone keep this up to date. Keep in mind this is not the price information set and used in Solar.web 🙄

As, at least in my system, I sometimes get Timeouts from the Datalogger (in their Status code "LN-timeout" whatever that means) I tried to keep the request volume low. 

My plan was to extend the integration with an option flow later so users can set individual update intervals for the devices. Then I could imagine to add logger info with a default interval of 1 hour - 1 day instead of 1 minute like the other endpoints.

farmio commented:
That said, if you think it's useful I can add it, no problem. I can just use a constant 1 hour scan interval for now...

nielstron commented:
> I deliberately left out Logger info for now because it only adds values that are constants which have to be set in the inverters Web-UI by the user. I think the least users have even set this, let alone keep this up to date. Keep in mind this is not the price information set and used in Solar.web roll_eyes

I see. The PR description just said that this would transfer *all* the sensors that were there before except for timestamp, which is simply incorrect right now.

I think that adding the logger is good, even if the user did not take care of it before it would be a good opportunity to remind the user of it. The upcoming versions of home assistant will provide options to monitor the price of energy where the information provided by logger info becomes especially interesting.

nielstron commented:
If the values provided are really only user-managed and not dynamically updated by the system I think 1 hour (or less) updates sound reasonable.

farmio committed:
prevent parallel requests to one host

farmio committed:
logger_info coordinator

farmio commented:
I have included a logger_coordinator now. 

The returned data from Fronius looks like this:
```json
"delivery_factor": {
    "value": 0.15000000596046448,
    "unit": "EUR/kWh"
},
```
Unfortunately there are some pitfalls:
- Fronius does a bad job at converting floats so we receive `0.15000000596046448` for what is set to `0.15` in their web UI.
- the unit "EUR/kWh" is not device_class "monetary" compatible - it should be only EUR (or whatever currency) for that I think. I used the returned unit for now and didn't set a device_class. We could change the unit in pyfronius if we wanted to have the device_class set.
- I have no idea where "co2_factor" comes from. I can't find it in the web ui of my device but get a value returned 🤷‍♂️

nielstron commented:
> Fronius does a bad job at converting floats so we receive 0.15000000596046448 for what is set to 0.15 in their web UI.

That is true. I recall that some rounding was implemented in the homeassistant fronius before because of the slight imprecision caused by fronius doubles.

> the unit "EUR/kWh" is not device_class "monetary" compatible - it should be only EUR (or whatever currency) for that I think. I used the returned unit for now and didn't set a device_class. We could change the unit in pyfronius if we wanted to have the device_class set.

when I tried last EUR/kWh could be handled by the energy dashboard... I didn't assign a special device class I think

farmio committed:
store FroniusSolarNet reference directly in coordinator

farmio committed:
cleanup coordinators when unloading entry

farmio commented:
> I recall that some rounding was implemented in the homeassistant fronius before

https://github.com/home-assistant/core/blob/30c25d44485cb3002d171f43049a5fa168bd2d40/homeassistant/components/fronius/sensor.py#L345

I think this is a bit unfortunate, because eg. grid tariffs can have more decimals than 2 - my export tariff for example is 0,076 €/kWh which would yield very different results when rounded to 2.
Should we do the rounding in pyfronius? There we could do it more granular and I think it fits since it is a Fronius issue.

nielstron commented:
The issue might stem from the underlying demjson to use the Decimal class rather than float to parse numbers. Could you try what happens if we use float instead of round? Otherwise, would 4 digits be a reasonable number of digits for any value?

farmio commented:
> the underlying demjson to use the Decimal class rather than float to parse numbers

Where is that used?

I think 4 digits sound reasonable. Will try how many decimals the web Ui even allows to configure there. 
Should we only do the rounding at LoggerSensor classes? All other classes values seem reasonable.

nielstron commented:
Whoops, mixed that up (its used in pysyncthru, not pyfronius). Checking the configurable precision in the Web UI is a good idea. And yeah, I agree that all other values seem good but I recall that this issue was present before the introduction of the Logger Info class.

farmio commented:
So the web-UI saves up to 3 decimals (and rounds if I enter more). 
`power_flow` also reports eg. `"P_Load" : -1209.5599999999995,` 
so I guess rounding everything to 4 digits would be ok.

farmio committed:
round floats returned by Fronius API

farmio committed:
default icons for grid im/export tariffs

farmio commented:
So I've added the rounding here, turned out to fit pretty nicely. 

@nielstron would you, as the codeowner, approve the overall changes or is something still missing / not appropriate?

nielstron reviewed (approved):
Looks good overall. Note: this will rename a number of entities and might break some customizations for previous users.

farmio commented:
That's right. I'll add a note to the PR description.

farmio committed:
small typing fix

janiversen commented:
you write "Since the integration now provides entities with unique_ids it can be changed back to the previous value if desired."

But why not do that automatically when you import the configuration.yaml ?

Bre77 commented on homeassistant/components/fronius/config_flow.py:
```diff
@@ -0,0 +1,114 @@
+            host=host,
+            is_logger=False,
+        )
+    raise CannotConnect("No supported Fronius SolarNet device found.")
```
You will need to add a test for this. config_flow.py needs 100% coverage.

farmio replied:
Thanks, didn't know that.

farmio commented:
Hi 👋!
The new entity_id look like this: `sensor.energy_total`
Vs. the old entity_ids looked like this:
```
sensor.energy_total_fronius_inverter_1_http_10_1_2_6
       |            |       |        | |_ url
       |            |       |        |_ solarnet id
       |            |       |_ device type
       |            |_ domain
       |_ key
```
Seeing this now I'm not particular happy with either. 
- the new ones are kind of sparse - they should probably at least contain the domain. Is there a common practice for this? maybe `sensor.fronius_inverter_1_energy_total`?
- it was a common (for a ~330 users) complaint in the community forums that the old style entity_ids were too long and cumbersome - therefore some users chose a custom component from HACS over the core component.
- the old style included the URL, which is probably not a good idea for the same reasons URLs are not applicable for unique_ids - if the URL changes the entity_id contains an IP address of a probably unrelated device
- the old ones practically contained every configuration key (device, sensor_type, scope) - these are not used anymore for meter and storage (system scope API endpoints are now always used instead of individual or system scope) so it would be a little cumbersome to recreate.

> But why not do that automatically when you import the configuration.yaml ?

We don't know which entities were used because they were (and still are) added dynamically as values appear in API responses and they have not had a unique_id so nothing entity_registry entry.

Can't really make up my mind about that. I'm open for suggestions.

Bre77 commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -1,354 +1,226 @@
+        super().__init__(coordinator)
+        self.entity_description = entity_description
+        self.solar_net_id = solar_net_id
+        self._attr_native_value = self._get_entity_value()
```
Does this work? My understanding is that this will set the native value on init only. I would expect you should just rename the `_get_entity_value` function to `native_value`.

farmio replied:
Famous last words: It works fine on my system 🤣

The entities are instantiated after instantiating the coordinator, so `_handle_coordinator_update()` is not called after an entity is created and would not have an init value then. Later updates are handled by `_handle_coordinator_update()`

Since we can't be sure if the key is contained in the latest API response (`coordinator.data` dict) the state would turn to `None` instead of stay on the latest known value. This happens eg. after sunset when an inverter doesn't produce anymore - Fronius API just omits some keys then 🙄

nielstron replied:
Actually, the state turning to None is not a problem. This is the intended way of telling HA that the current state could not be fetched. AFAIK HA will still display the latest non-None retrieved value.

farmio replied:
I've just tested this. At night I would get these API results from inverters - alternating:
```log
[pyfronius] Converting inverter data from '{'DAY_ENERGY': {'Unit': 'Wh', 'Value': 112640}, 'DeviceStatus': {'ErrorCode': 307, 'LEDColor': 1, 'LEDState': 0, 'MgmtTimerRemainingTime': 14, 'StateToReset': False, 'StatusCode': 3}, 'IDC': {'Unit': 'A', 'Value': 0}, 'TOTAL_ENERGY': {'Unit': 'Wh', 'Value': 43421500}, 'UDC': {'Unit': 'V', 'Value': 4.4}, 'YEAR_ENERGY': {'Unit': 'Wh', 'Value': 24742314}}'
[pyfronius] Converting inverter data from '{'DAY_ENERGY': {'Unit': 'Wh', 'Value': 112640}, 'DeviceStatus': {'ErrorCode': 523, 'LEDColor': 3, 'LEDState': 0, 'MgmtTimerRemainingTime': -1, 'StateToReset': True, 'StatusCode': 2}, 'TOTAL_ENERGY': {'Unit': 'Wh', 'Value': 43421500}, 'YEAR_ENERGY': {'Unit': 'Wh', 'Value': 24742314}}'
```
ErrorCode `307` contains Current and Voltage, `523` does not. 
With a `native_value` property returning `None` if the key was not in the result the state would alternate between `Unknown` and `0` every minute or so (its not regularly alternating).

Bre77 replied:
Well I figured out where I was misunderstanding this, you are replacing the `_handle_coordinator_update` callback to update `_attr_native_value`.

Bre77 commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -1,354 +1,226 @@
+        """Extract information for SolarNet device from coordinator data."""
+        return self.coordinator.data[self.solar_net_id]  # type: ignore[no-any-return]
+
+    def _get_entity_value(self) -> Any:
```
As above, suggest this be called `native_value` so its used directly.

Bre77 reviewed (changes requested):
This is a big change. I believe you might need to improve test coverage of your config flow and I am unsure the way you have implemented the sensor native value is correct.

jmtatsch commented:
I agree that the old naming scheme is very clunky and not really consistent with other integrations. 
Why not introduce a name when setting up the config flow (default can be the url) and then do
my_name_energy_total

farmio commented:
I like the idea of a configurable name, but I'm not sure if adding it to the entity_id is a good idea because changing the name later would unexpectedly either change or not change the associated entity_ids (depending on a users expectations which I guess vary 🙃 in that case).

Bre77 commented:
@farmio it would certainly make sense for the display name to use the name configured in **Fronius** itself, but the unique ID could be prefixed with the host. I unfortuantely cannot find any of the APIs that give the name in Fronius though.

Bre77 commented on homeassistant/components/fronius/sensor.py:
```diff
@@ -1,354 +1,226 @@
+        meter_data = self._device_data()
+
+        self._attr_extra_state_attributes = {
+            "meter_loaction": meter_data["meter_location"]["value"],
```
Typo, should be **meter_location**

Bre77 replied:
```suggestion
            "meter_location": meter_data["meter_location"]["value"],
```

farmio committed:
Update homeassistant/components/fronius/sensor.py

Co-authored-by: Brett Adams <Bre77@users.noreply.github.com>

frenck commented:
This PR needs to be broken down into multiple PRs. We should not accept big overhauls like this, covering multiple feature changes, in a single PR.

farmio committed:
Merge branch 'fronius-refactor' of https://github.com/farmio/home-assistant into fronius-refactor

farmio committed:
DC icons

farmio committed:
Merge remote-tracking branch 'upstream/dev' into fronius-refactor

farmio committed:
prepend names with "Fronius" and device type

to get more reasonable default entity_ids (eg. have them next to each other when alphabetically sorted)

farmio committed:
Merge remote-tracking branch 'upstream/dev' into fronius-refactor

farmio commented:
Well there is a version now that only includes the move to DataUpdateCoordinator and addition of unique_id #57879

farmio commented:
closed in favor of #59686
